File: global-hooks/framework/automation/file_watcher.py
#!/usr/bin/env python3
"""FileChanged Hook — watches dependency files (package.json, pyproject.toml,
Cargo.toml, go.mod). Logs changes and injects advisory context. Exit: always 0."""
import json, sys
from datetime import datetime, timezone
from pathlib import Path

WATCHED = {"package.json", "pyproject.toml", "Cargo.toml", "go.mod"}
LOG_PATH = Path.home() / ".claude" / "data" / "file_changes.jsonl"


def main():
    try:
        data = json.load(sys.stdin)
        session_id = data.get("session_id", "unknown")
        file_path = data.get("file", {}).get("filePath", "") or data.get("filePath", "")
        filename = Path(file_path).name if file_path else ""
        if filename not in WATCHED:
            print(json.dumps({}))
            sys.exit(0)
        # Log to JSONL
        LOG_PATH.parent.mkdir(parents=True, exist_ok=True)
        entry = {"timestamp": datetime.now(timezone.utc).isoformat(),
                 "session_id": session_id, "file": file_path, "filename": filename}
        with open(LOG_PATH, "a") as f:
            f.write(json.dumps(entry) + "\n")
        # Inject advisory context
        output = {"hookSpecificOutput": {
            "hookEventName": data.get("hook_event_name", "FileChanged"),
            "additionalContext": (
                f"Dependency file {filename} was modified externally. "
                "Consider running dependency audit or checking for breaking changes."
            ),
        }}
        print(json.dumps(output))
    except Exception as e:
        print(f"file_watcher error (non-blocking): {e}", file=sys.stderr)
    sys.exit(0)


if __name__ == "__main__":
    try:
        main()
    except Exception:
        pass

File: caf-hooks/src/hooks/file_watcher.rs
/// FileChanged hook — watches dependency files.
///
/// Python equivalent: global-hooks/framework/automation/file_watcher.py (45 LOC)
///
/// Behavior:
/// - Only process: package.json, pyproject.toml, Cargo.toml, go.mod
/// - Log change to ~/.claude/data/file_changes.jsonl
/// - Inject advisory context
/// - Always exits 0
use chrono::Utc;
use serde::Serialize;
use serde_json::Value;
use std::path::Path;

use crate::io::{read_stdin_value, try_append_jsonl, write_output};
use crate::state::file_changes_path;
use crate::types::HookOutput;

const WATCHED: &[&str] = &["package.json", "pyproject.toml", "Cargo.toml", "go.mod"];

#[derive(Serialize)]
struct FileChangeRecord<'a> {
    timestamp: String,
    session_id: &'a str,
    file: &'a str,
    filename: &'a str,
}

pub fn run() {
    let data: Value = read_stdin_value();

    let session_id = data.get("session_id")
        .and_then(|v| v.as_str())
        .unwrap_or("unknown")
        .to_string();

    let hook_event = data.get("hook_event_name")
        .and_then(|v| v.as_str())
        .unwrap_or("FileChanged")
        .to_string();

    // file_path can come from data["file"]["filePath"] or data["filePath"]
    let file_path = data
        .get("file")
        .and_then(|f| f.get("filePath"))
        .and_then(|v| v.as_str())
        .or_else(|| data.get("filePath").and_then(|v| v.as_str()))
        .or_else(|| data.get("file_path").and_then(|v| v.as_str()))
        .unwrap_or("")
        .to_string();

    let filename = Path::new(&file_path)
        .file_name()
        .and_then(|n| n.to_str())
        .unwrap_or("")
        .to_string();

    // Only process watched files
    if !WATCHED.contains(&filename.as_str()) {
        write_output(&serde_json::json!({}));
        return;
    }

    // Log to JSONL
    let record = FileChangeRecord {
        timestamp: Utc::now().to_rfc3339(),
        session_id: &session_id,
        file: &file_path,
        filename: &filename,
    };
    try_append_jsonl(&file_changes_path(), &record);

    // Inject advisory context
    let msg = format!(
        "Dependency file {} was modified externally. \
        Consider running dependency audit or checking for breaking changes.",
        filename
    );
    let output = HookOutput::inject_context(hook_event, msg);
    write_output(&output);
}
